authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,583 @@
+                LOG.warn("Revoke role: {} from group: {} failed!", role, group, e);
+              }
+            });
+
+    return Boolean.TRUE;
+  }
+
+  @Override
+  public Boolean onUserAdded(User user) throws RuntimeException {
+    VXUserList list = rangerClient.searchUser(ImmutableMap.of("name", user.name()));
+    if (list.getListSize() > 0) {
+      LOG.warn("The user({}) already exists in the Ranger!", user.name());
+      return Boolean.FALSE;
+    }
+
+    VXUser rangerUser = VXUser.builder().withName(user.name()).withDescription(user.name()).build();
+    return rangerClient.createUser(rangerUser);
+  }
+
+  @Override
+  public Boolean onUserRemoved(User user) throws RuntimeException {
+    VXUserList list = rangerClient.searchUser(ImmutableMap.of("name", user.name()));
+    if (list.getListSize() == 0) {
+      LOG.warn("The user({}) doesn't exist in the Ranger!", user);
+      return Boolean.FALSE;
+    }
+    rangerClient.deleteUser(list.getList().get(0).getId());
+    return Boolean.TRUE;
+  }
+
+  @Override
+  public Boolean onUserAcquired(User user) throws RuntimeException {
+    VXUserList list = rangerClient.searchUser(ImmutableMap.of("name", user.name()));
+    if (list.getListSize() == 0) {
+      LOG.warn("The user({}) doesn't exist in the Ranger!", user);
+      return Boolean.FALSE;
+    }
+    return Boolean.TRUE;
+  }
+
+  @Override
+  public Boolean onGroupAdded(Group group) throws RuntimeException {
+    return rangerClient.createGroup(VXGroup.builder().withName(group.name()).build());
+  }
+
+  @Override
+  public Boolean onGroupRemoved(Group group) throws RuntimeException {
+    VXGroupList list = rangerClient.searchGroup(ImmutableMap.of("name", group.name()));
+    if (list.getListSize() == 0) {
+      LOG.warn("The group({}) doesn't exist in the Ranger!", group);
+      return Boolean.FALSE;
+    }
+    return rangerClient.deleteGroup(list.getList().get(0).getId());
+  }
+
+  @Override
+  public Boolean onGroupAcquired(Group group) {
+    VXGroupList vxGroupList = rangerClient.searchGroup(ImmutableMap.of("name", group.name()));
+    if (vxGroupList.getListSize() == 0) {
+      LOG.warn("The group({}) doesn't exist in the Ranger!", group);
+      return Boolean.FALSE;
+    }
+    return Boolean.TRUE;
+  }
+
+  /**
+   * Add the securable object's privilege to the Ranger policy. <br>
+   * 1. Find the policy base the metadata object. <br>
+   * 2. If the policy exists and has the same privilege, because support idempotent operation, so
+   * return true. <br>
+   * 3. If the policy does not exist, then create a new policy. <br>
+   */
+  private boolean doAddSecurableObject(RoleChange.AddSecurableObject change) {
+    RangerPolicy policy = rangerHelper.findManagedPolicy(change.getSecurableObject());
+
+    if (policy != null) {
+      // Check the policy item's accesses and roles equal the Gravitino securable object's privilege
+      Set<String> policyPrivileges =
```

Comment:
Could we reuse the method mappingPrivilege here?

Reply:
They have different logic